src/EditorFeatures/Core.Wpf/InheritanceMargin/InheritanceMarginTag.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.InheritanceMargin;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Text.Editor;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InheritanceMargin
+{
+    internal class InheritanceMarginTag : IGlyphTag
+    {
+        /// <summary>
+        /// Margin moniker.
+        /// </summary>
+        public readonly ImageMoniker Moniker;
+
+        /// <summary>
+        /// Members needs to be shown on this line. There might be multiple members.
+        /// For example:
+        /// interface IBar { void Foo1(); void Foo2() }
+        /// class Bar : IBar { void Foo1() { } void Foo2() { } }
+        /// </summary>
+        public readonly ImmutableArray<InheritanceMemberItem> MembersOnLine;
+
+        public InheritanceMarginTag(ImmutableArray<InheritanceMemberItem> membersOnLine)
+        {
+            Debug.Assert(!membersOnLine.IsEmpty);
+
+            MembersOnLine = membersOnLine;
+            var aggregateRelationship = membersOnLine
+                .SelectMany(member => member.TargetItems.Select(target => target.RelationToMember))
+                .Aggregate((r1, r2) => r1 | r2);
```

Comment:
i don't love that the common case (a single item) will incur all these allocations all over the place.

Reply:
Seperate those cases